### PR TITLE
Add chapters on large scale data.

### PR DESCRIPTION
**Pull request by dmbates**

### CODE REVIEW

**Comment by dmbates** [comment]
`largescaleobserved.qmd` uses `./data/ratings.arrow` which is not in the repository as it is much larger (164 MB) than github allows.  I think one way to address the issue of the large arrow files required in chapters 4 and 5 is to make the repository both a Julia package and a Quarto project and to incorporate the current ml-25m package into this package.  I don't think either package needs to be registered but the code to construct the data sets should be available in a repository.

**Comment by palday** [comment]
> I think one way to address the issue of the large arrow files required in chapters 4 and 5 is to make the repository both a Julia package and a Quarto project and to incorporate the current ml-25m package into this package. I don't think either package needs to be registered but the code to construct the data sets should be available in a repository.

Sounds good

**Comment by dmbates** [comment]
@palday Are you waiting for me to do something before you review this PR?

**Comment by palday** [comment]
IIRC I ran into some snag when trying to render the new chapters locally and then got busy and haven't had a chance to revisit. Sorry! 🙂

**Comment by dmbates** [comment]
> IIRC I ran into some snag when trying to render the new chapters locally and then got busy and haven't had a chance to revisit. Sorry! slightly_smiling_face

`largescaleobservational.qmd` requires the ml-latest download and conversion of the ratings to arrow format.  The file itself is too big for github.  Would you like me to make this into a Julia project and copy over the code from the ml-25m repository so that that ml-latest.zip file can be downloaded if needed?

File: largescaleobserved.qmd [hunk 1]
```diff
@@ -0,0 +1,508 @@
...
+  :userId => (x -> Int32(length(x))) => :urtngs,
+)
+extrema(users.urtngs)  # minimum and maximum number of ratings per user
+```
+
+One way of visualizing the imbalance in the number of ratings per movie or per user is as an [empirical cumulative distribution function](https://en.wikipedia.org/wiki/Empirical_distribution_function) (ecdf) plot, which is a "stair-step" plot where the vertical axis is the proportion of observations less than or equal to the corresponding value on the horizontal axis.
+Because the distribution of the number of ratings per movie or per user is so highly skewed in the low range we use a logarithmic horizontal axis in @fig-nrtngsecdf.
+
+```{julia}
+#| code-fold: true
```
Comment (palday) [inline largescaleobserved.qmd:116]: I get a weird error on this block:
```julia
ERROR: ArgumentError: range(1.0, stop=81491.0, length=1): endpoints differ
Stacktrace:
  [1] _linspace1(#unused#::Type{Float64}, start::Float64, stop::Float64, len::Int64)
    @ Base ./twiceprecision.jl:757
  [2] _linspace(#unused#::Type{Float64}, start_n::Int32, stop_n::Int32, len::Int32, den::Int32)
    @ Base ./twiceprecision.jl:737
  [3] _linspace
    @ ./twiceprecision.jl:734 [inlined]
  [4] range_start_stop_length
    @ ./range.jl:571 [inlined]
  [5] _range
    @ ./range.jl:157 [inlined]
  [6] #range#69
    @ ./range.jl:140 [inlined]
  [7] ecdf_xvalues(ecdf::StatsBase.ECDF{Vector{Int32}, StatsBase.Weights{Float64, Float64, Vector{Float64}}}, npoints::Int32)
    @ Makie ~/.julia/packages/Makie/Ppzqh/src/stats/ecdf.jl:5
  [8] convert_arguments(P::Type{Combined{Makie.stairs}}, ecdf::StatsBase.ECDF{Vector{Int32}, StatsBase.Weights{Float64, Float64, Vector{Float64}}}; npoints::Int32)
    @ Makie ~/.julia/packages/Makie/Ppzqh/src/stats/ecdf.jl:11
  [9] #convert_arguments#773
    @ ~/.julia/packages/Makie/Ppzqh/src/stats/ecdf.jl:61 [inlined]
 [10] plot!(scene::Scene, P::Type{Combined{Makie.ecdfplot}}, attributes::Attributes, args::Vector{Int32}; kw_attributes::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ Makie ~/.julia/packages/Makie/Ppzqh/src/interfaces.jl:317
 [11] plot!(scene::Scene, P::Type{Combined{Makie.ecdfplot}}, attributes::Attributes, args::Vector{Int32})
    @ Makie ~/.julia/packages/Makie/Ppzqh/src/interfaces.jl:302
 [12] plot(P::Type{Combined{Makie.ecdfplot}}, gp::GridPosition, args::Vector{Int32}; axis::NamedTuple{(:xlabel, :xminorgridvisible, :xminorticks, :xminorticksvisible, :xscale, :ylabel, :yminorticksvisible), Tuple{String, Bool, IntervalsBetween, Bool, typeof(log10), String, Bool}}, kwargs::Base.Pairs{Symbol, Int32, Tuple{Symbol}, NamedTuple{(:npoints,), Tuple{Int32}}})
    @ Makie ~/.julia/packages/Makie/Ppzqh/src/figureplotting.jl:95
 [13] ecdfplot(::GridPosition, ::Vararg{Any}; attributes::Base.Pairs{Symbol, Any, Tuple{Symbol, Symbol}, NamedTuple{(:npoints, :axis), Tuple{Int32, NamedTuple{(:xlabel, :xminorgridvisible, :xminorticks, :xminorticksvisible, :xscale, :ylabel, :yminorticksvisible), Tuple{String, Bool, IntervalsBetween, Bool, typeof(log10), String, Bool}}}}})
    @ Makie ~/.julia/packages/MakieCore/rLlRw/src/recipes.jl:34
 [14] top-level scope
    @ code:9
```

File: largescaleobserved.qmd [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+@fig-timetoconverge shows that the average evaluation time for the log-likelihood function depends strongly on the number of movies and less strongly on the number of users.
+
+However the middle panel shows that the number of iterations to convergence is highly variable.
+Most of these models required between 20 and 25 evaluations but some required almost 50 evaluations.
+
+The derivation of the log-likelihood for linear mixed-effects models is given in @sec-lmmtheory, which provides a rather remarkable result:
+the profiled log-likelihood for a linear mixed-effects model can be evaluated from Cholesky factor of a blocked, positive-definite symmetric matrix.
+
+There are two blocked matrices, `A` and `L`, stored in the model representation and, for large models such as we are considering these are the largest fields in 
+## Model size and speed for different thresholds
```
Comment (palday) [inline largescaleobserved.qmd:489]:
```suggestion
There are two blocked matrices, `A` and `L`, stored in the model representation and, for large models such as we are considering these are the largest fields in 

## Model size and speed for different thresholds
```

and there seems to be something missing after `in`.

File: largescaleobserved.qmd (hunk 1 above)
Comment (palday) [reply largescaleobserved.qmd:116]: I still had to do some tinkering -- I think my local copy of the big arrow table differs a little from yours, but this is something we can fix with a dedicated data package.

**Review by palday — APPROVED** [review]
Looks good! It's got me really excited about RFP storage.

**Comment by palday** [comment]
suggestion to speed along review for these things -- let's do the source changes in one PR (which I always render locally to make sure things are reproducible) and then the updates to the rendered output in a follow-up PR. This would make the PR where I'm actually looking at diff's more succinct. :smile:

**Comment by dmbates** [comment]
Sounds good.  I will split the PR's as you suggest.

**Pull request merged**